File: crypto-core/ts/src/message.ts
// Change Message M(c) and the on-chain field record. ALGORITHMS.md §1.1–§1.2.
import { Writer, Reader, utf8 } from "./bytes.js";

export const MAGIC_M = utf8("TEMC"); // Triple-Entry Message Canonical
export const MAGIC_R = utf8("TER1"); // Triple-Entry Record v1
export const VERSION = 1;

export enum Op {
  INSERT = 1,
  UPDATE = 2,
  DELETE = 3,
}

export enum ImageKind {
  PLAINTEXT = 0,
  COMMITMENT = 1,
}

export interface ChangeMessage {
  tableId: string;
  rowId: Uint8Array;
  columnId: string;
  op: Op;
  seq: bigint;
  prevTxid: Uint8Array; // 32 bytes, or empty at genesis
}

/** encode M(c) — the public change identity hashed to GV. ALGORITHMS.md §1.1 */
export function encodeMessage(m: ChangeMessage): Uint8Array {
  if (m.prevTxid.length !== 0 && m.prevTxid.length !== 32) throw new Error("prevTxid must be empty or 32 bytes");
  return new Writer()
    .raw(MAGIC_M)
    .u8(VERSION)
    .str(m.tableId)
    .bytes(m.rowId)
    .str(m.columnId)
    .u8(m.op)
    .u64(m.seq)
    .bytes(m.prevTxid)
    .finish();
}

export interface FieldRecord {
  streamId: Uint8Array;
  seq: bigint;
  prevTxid: Uint8Array;
  imageKind: ImageKind;
  changeImage: Uint8Array;
  tag: Uint8Array; // 32 bytes
}

/** encode the on-chain field record carried as spendable-script pushdata. ALGORITHMS.md §1.2 */
export function encodeRecord(r: FieldRecord): Uint8Array {
  if (r.prevTxid.length !== 0 && r.prevTxid.length !== 32) throw new Error("prevTxid must be empty or 32 bytes");
  if (r.tag.length !== 32) throw new Error("tag must be 32 bytes");
  return new Writer()
    .raw(MAGIC_R)
    .u8(VERSION)
    .bytes(r.streamId)
    .u64(r.seq)
    .bytes(r.prevTxid)
    .u8(r.imageKind)
    .bytes(r.changeImage)
    .bytes(r.tag)
    .finish();
}

export function decodeRecord(buf: Uint8Array): FieldRecord {
  const rd = new Reader(buf);
  const magic = rd.raw(4);
  for (let i = 0; i < 4; i++) if (magic[i] !== MAGIC_R[i]) throw new Error("bad record magic");
  const version = rd.u8();
  if (version !== VERSION) throw new Error(`unsupported record version ${version}`);
  const streamId = rd.bytes();
  const seq = rd.u64();
  const prevTxid = rd.bytes();
  const imageKind = rd.u8() as ImageKind;
  if (imageKind !== ImageKind.PLAINTEXT && imageKind !== ImageKind.COMMITMENT) throw new Error("bad image_kind");
  const changeImage = rd.bytes();
  const tag = rd.bytes();
  rd.end();
  if (tag.length !== 32) throw new Error("tag must be 32 bytes");
  return { streamId, seq, prevTxid, imageKind, changeImage, tag };
}

File: crypto-core/go/message.go
package cryptocore

import "errors"

var (
	MagicM  = []byte("TEMC") // Triple-Entry Message Canonical
	MagicR  = []byte("TER1") // Triple-Entry Record v1
	Version = byte(1)
)

type Op uint8

const (
	OpInsert Op = 1
	OpUpdate Op = 2
	OpDelete Op = 3
)

type ImageKind uint8

const (
	ImagePlaintext  ImageKind = 0
	ImageCommitment ImageKind = 1
)

// ChangeMessage is the public change identity M(c) (ALGORITHMS.md §1.1).
type ChangeMessage struct {
	TableID  string
	RowID    []byte
	ColumnID string
	Op       Op
	Seq      uint64
	PrevTxid []byte // 32 bytes, or empty at genesis
}

// EncodeMessage encodes M(c). ALGORITHMS.md §1.1.
func EncodeMessage(m ChangeMessage) ([]byte, error) {
	if len(m.PrevTxid) != 0 && len(m.PrevTxid) != 32 {
		return nil, errors.New("prevTxid must be empty or 32 bytes")
	}
	return NewWriter().
		Raw(MagicM).
		U8(Version).
		Str(m.TableID).
		Bytes(m.RowID).
		Str(m.ColumnID).
		U8(byte(m.Op)).
		U64(m.Seq).
		Bytes(m.PrevTxid).
		Finish(), nil
}

// FieldRecord is the on-chain field record carried as spendable-script pushdata (ALGORITHMS.md §1.2).
type FieldRecord struct {
	StreamID    []byte
	Seq         uint64
	PrevTxid    []byte
	ImageKind   ImageKind
	ChangeImage []byte
	Tag         []byte // 32 bytes
}

func EncodeRecord(r FieldRecord) ([]byte, error) {
	if len(r.PrevTxid) != 0 && len(r.PrevTxid) != 32 {
		return nil, errors.New("prevTxid must be empty or 32 bytes")
	}
	if len(r.Tag) != 32 {
		return nil, errors.New("tag must be 32 bytes")
	}
	return NewWriter().
		Raw(MagicR).
		U8(Version).
		Bytes(r.StreamID).
		U64(r.Seq).
		Bytes(r.PrevTxid).
		U8(byte(r.ImageKind)).
		Bytes(r.ChangeImage).
		Bytes(r.Tag).
		Finish(), nil
}

func DecodeRecord(buf []byte) (FieldRecord, error) {
	var rec FieldRecord
	rd := NewReader(buf)
	magic := rd.Raw(4)
	if rd.Err() == nil && string(magic) != string(MagicR) {
		return rec, errors.New("bad record magic")
	}
	if v := rd.U8(); rd.Err() == nil && v != Version {
		return rec, errors.New("unsupported record version")
	}
	rec.StreamID = rd.Bytes()
	rec.Seq = rd.U64()
	rec.PrevTxid = rd.Bytes()
	kind := rd.U8()
	if rd.Err() == nil && kind != byte(ImagePlaintext) && kind != byte(ImageCommitment) {
		return rec, errors.New("bad image_kind")
	}
	rec.ImageKind = ImageKind(kind)
	rec.ChangeImage = rd.Bytes()
	rec.Tag = rd.Bytes()
	if err := rd.End(); err != nil {
		return rec, err
	}
	if len(rec.Tag) != 32 {
		return rec, errors.New("tag must be 32 bytes")
	}
	return rec, nil
}
